HTML API: Parse doctypes and set full parser quirks mode correctly

sirreal opened the pull request:
- Add complete parsing of DOCTYPE tokens. This is handled by a new class `WP_HTML_Doctype_Info` that handles parsing DOCTYPE tokens and exposes information about them.
- Pause on DOCTYPE tokens in initial mode from the HTML processor. These tokens are not ignored (like most doctype tokens) and pausing allows the doctype to be visited and inspected by consumers of the HTML API.
- Update full parser document mode (quirks mode) according to doctype tokens [as described in the specification](https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode).
- When the full parser advances from initial mode in the "any other" condition, it should set quirks mode. This has been added.
- Update the html5lib tests to handle doctypes.

46 skipped tests from HTML5lib are now run. 1 test was disabled due to the way some whitespace is handled in the full parser.

```diff
 OK, but incomplete, skipped, or risky tests!
-Tests: 1498, Assertions: 930, Skipped: 568.
+Tests: 1497, Assertions: 975, Skipped: 522.
```

---

This change adds a new class to handle DOCTYPE token information according to the specification. The class is exposed from Tag and HTML processors when a DOCTYPE token is reached. DOCTYPE token information can be retrieved for inspection by calling `$processor->get_doctype_info();`. See this example form the HTML5lib-tests:

https://github.com/WordPress/wordpress-develop/blob/2ffff8d3c45539bb2770fc1372ccc243d877f3e5/tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php#L192-L200

The new class parses DOCTYPE Tokens in greater detail. This is useful because DOCTYPE tokens may appear in many places in HTML but are ignores in most situations. The detailed parsing of DOCTYPE tokens to be handled on-demand when a DOCTYPE token is reached under the appropriate circumstances.

The `WP_HTML_Doctype_Info` class also handles the complex rules for determining quirks mode which involve inspecting the DOCTYPE token name, public identifier, system identifier, and force_quirks_flag.

Trac ticket: https://core.trac.wordpress.org/ticket/61576

---

<details><summary>Survey of existing DOCTYPE declarations</summary>

Download [the DOCTYPE report](https://github.com/user-attachments/files/16639422/report-doctypes.txt) and `cat report-doctypes.txt` to see the color output.

Here is a preview:

<img width="1287" alt="Screenshot 2024-08-16 at 9 46 17 AM" src="https://github.com/user-attachments/assets/6417f4b8-b295-421f-a496-ab7a00f35bb5">

<img width="1122" alt="Screenshot 2024-08-16 at 9 46 31 AM" src="https://github.com/user-attachments/assets/84e6e661-0efd-4ea3-b291-87318376c97e">

</details>

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
Stop on DOCTYPE tokens in next_token

Similar to text nodes, this change adds DOCTYPE tokens to the stack of
open elements so they can be reached when stepping through the document
via `next_token`.

sirreal committed:
Handle doctype tokens in html5lib-tests

sirreal committed:
Add get_doctype_name method to tag processor

This method handles parsing the doctype name from a doctype declaration.
This is important for the full HTML processor to be able to correctly
determin whether it is in quirks mode.

sirreal committed:
Add missing-whitespace-before-doctype-name test

sirreal committed:
Allow gotos in tag processor

sirreal committed:
Add get_compat_mode method

sirreal committed:
WIP parsing doctypes

sirreal committed:
Handle system doctype

sirreal committed:
Handle public-id only + whitespace

sirreal commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -4026,6 +4026,146 @@ private function matches(): bool {
+		return $doctype[0];
+	}
+
+	public function parse_doctype(): ?array {
```
I think this should be protected, not public. We can set doctype details on processor state and add getters for them. Then we can do this parsing just once when the doctype is reached.

Maybe there's a better place for this parsing to happen.

sirreal commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -4026,6 +4026,146 @@ private function matches(): bool {
+		 * https://html.spec.whatwg.org/multipage/parsing.html#before-doctype-name-state
+		 */
+
+
```
```suggestion
```

sirreal committed:
lint

sirreal committed:
Update html5lib-tests to get doctype name, publicid, systemid

sirreal committed:
Parsing doctypes and handling quirks mode correctly

There are some discrepancies with browser behavior to work out.

sirreal committed:
Fix logic error when parsing public and system identifiers

sirreal committed:
Disable pre-body failing whitespace text test

sirreal committed:
Fix multiline quote

sirreal committed:
Scaffold doctype info class

sirreal committed:
Return DOCTYPE info class from get_doctype_info method

sirreal committed:
Move quirks detect to get_compatibility_mode method

sirreal committed:
Remove get_compat_mode method from processor class

sirreal committed:
Always return string on doctype attribute lookups

sirreal committed:
Update tests to use get_doctype_info function

sirreal committed:
Update tests to use doctype info

sirreal commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -2939,4 +2939,45 @@ public function test_unclosed_funky_comment_input_too_short() {
+	 *
+	 * @dataProvider data_doctypes
+	 *
+	 * @ticket TBD
```
```suggestion
	 * @ticket 61576
```

sirreal committed:
Update test ticket number

sirreal commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -2939,4 +2939,45 @@ public function test_unclosed_funky_comment_input_too_short() {
+	 *
+	 * @ticket 61576
+	 */
+	public function test_doctype_doc_name( string $html, ?string $expected_doctype_name ) {
```
There are good doctype tests as part of the HTML5lib suite and they caught a few issues I had.

We can consider adding more tests here or some specific tests for the behavior of the doctype_info class, especially regarding get_compatibility_mode although even that is covered by the HTML5lib tests where they often use `<p><table>` as a test for whether quirks mode is set correctly.

sirreal committed:
Add "quirks mode" to "anything else" initial mode

> Anything else
> … set the Document to quirks mode.

dmsnell commented on src/wp-includes/html-api/class-wp-html-doctype-info.php:
```diff
@@ -0,0 +1,269 @@
+
+	/**
+	 * The system identifier of the DOCTYPE.
+	 *
```
I'm just guessing but I bet a lot could be added to these properties and this class to educate people on the value (or danger) of these parts of the DOCTYPE. Just a comment saying something like "This exists mostly to be able to represent and copy HTML from one document from another, but it's best to use only the standard `<!DOCTYPE html>` declaration for HTML."

just doodling

dmsnell commented on src/wp-includes/html-api/class-wp-html-doctype-info.php:
```diff
@@ -0,0 +1,269 @@
+	 *
+	 * @var bool
+	 */
+	private $force_quirks_flag;
```
perhaps `$forces_quirks_flag` might align better with the nuance? _this declaration forces a document into quirks mode_

sirreal replied:
I'd prefer to keep this `$force_quirks_flag` which aligns with the spec.

sirreal replied:
I've re-read some of your other related comments and I think I agree after all. I'm going to re-work some of this to establish the resulting quirks mode on construction instead of re-running the algorithm in the `get_quirks_mode` function.

dmsnell commented on src/wp-includes/html-api/class-wp-html-doctype-info.php:
```diff
@@ -0,0 +1,269 @@
+		?string $public_identifier,
+		?string $system_identifier,
+		bool $force_quirks_flag
+	) {
```
I don't love this constructor even though it's probably fine with `@access private`

Did you consider moving the parsing code inside of it? We could replace this with a static creator method like `createFromString( $html, $at, $length )` and parse evereything within this class.

dmsnell replied:
moving the parse of the token into this class could help resolve the `force-quirks-flag` information visibility problem as well.

when constructing, as soon as `force-quirks-flag` is activated, we can stop parsing and set `$indicates_document_mode = 'quirks';` and so the four properties can be fully: name, public identifier, system identifier, indicated document mode.

sirreal replied:
I moved the parsing into the constructor, made it access-private, and added some notes that it should really be constructed from a doctype token's `get_modifiable_text`.

The parser now expects what's "in" the doc type, and assumes that whatever it is passed is correct.

These classes become quite closely coupled. That may OK at this time given the narrow usefulness of this class (it's practically internal) and the documentation.

Curious to hear your thoughts.

dmsnell replied:
instead of tightly coupling these and recreating some of the pains that Core already has, such as the `$attr attribute list from HTML` in `wp_kses_hair()`, I think it would be worth it to double-parse the DOCTYPE tag in its entirety. We can pass the full token into the class and reparse, as parsing the DOCTYPE declaration is much easier than parsing a tag name, and should only happen at most once during the parse of a document, even if multiple tokens appear (since we can ignore the successive tokens).

sirreal replied:
I _almost_ passed the full token and then changed direction after looking at `get_modifiable_text`.

- get_modifiable_text handles newline normalization and null bytes. This isn't a big deal to duplicate.
- If there were pending updates from `set_modifiable_text`, we'd miss them because we'd be getting the "raw" token.

Thinking about that more, I think it's OK to miss updates because I don't think there should be any updates. Changing the doctype seems very dangerous because it has the potential to alter the document's compat mode which could affect how it's parsed.

I do have concerns about the failure mode. What happens if we don't get a valid doctype string? We can return a quirks mode doctype instance with all empty strings, but it would be nice to indicate somehow that the provided string wasn't a valid doctype. Would it be OK to throw an exception in that case given this constructor is `@access private` and we only expect to ever receive a valid DOCTYPE string from one of the processor classes?

dmsnell replied:
thankfully we don't currently support setting modifiable text for doctype declarations.

dmsnell replied:
> I do have concerns about the failure mode

it seems like if we can't parse it then we can't make any judgement about the document mode because it's not a real DOCTYPE declaration token. returning `null` and ignoring it seems fitting.

dmsnell commented on src/wp-includes/html-api/class-wp-html-doctype-info.php:
```diff
@@ -0,0 +1,269 @@
+	 */
+	public function get_name(): string {
+		return $this->name ?? '';
+	}
```
given that this basically serves as a record class with an intelligent constructor, do you see much value in hiding these behind `private`?

what if the properties were all `public` and set upon instantiation, including empty strings when not set?

sirreal replied:
I don't think it's OK to set these properties, that's one reason to use getters.

sirreal replied:
There's also the `null/""` difference in some of the quirks mode rules, but that's less relevant since I extracted the quirks mode algorithm to another method.

dmsnell replied:
it's not okay to set them, but also seems needless to recreate state-oriented interfaces that do nothing. I know "once public always public" occurs here, but I also see these as being closed to extension or updates.

we can also document them saying "don't change these" but I don't know why someone _would_ when the class doesn't do anything else. it doesn't support serialization and it shouldn't be re-parsing itself to change the suggested document mode.

in essence, they are read only unless someone changes them and re-reads them, and I'm not sure what the point of that would be.

sirreal committed:
Move doctype contents parsing into doctype_info

sirreal committed:
Better comments and naming

sirreal committed:
Improve more documentation in comments

sirreal committed:
Add more information to the class doc block

sirreal committed:
Determing compat mode on initial doctype parse

sirreal commented on src/wp-includes/html-api/class-wp-html-doctype-info.php:
```diff
@@ -0,0 +1,472 @@
+	private $system_identifier = null;
+
+	/**
+	 * The compatibility (quirks) mode of the document that results from parsing this DOCTYPE.
```
```suggestion
	 * The compatibility (quirks) mode of the document that results from parsing this DOCTYPE.
	 * One of "no-quirks", "limited-quirks", or "quirks".
```

sirreal committed:
Add more info about compatibility mode property strings

dmsnell commented on src/wp-includes/html-api/class-wp-html-doctype-info.php:
```diff
@@ -0,0 +1,473 @@
+	 *
+	 * @return string The name of the DOCTYPE.
+	 */
+	public function get_name(): string {
```
I almost think it would be more valuable to represent the absence of these identifiers, just as `get_attribute()` does. the doctype declaration can have an empty name and it can lack a name, I guess. maybe this is more about the system and public identifier.

if I were writing a method to copy the token from a source document into a normalized output, however, I would want to know the difference.

I also find it unlikely that userspace code is going to be doing much more with this than possibly copying HTML. I would like to see what people want or expect to do with the information, but it's not like we see people parsing DOCTYPE tokens already using the existing tools available

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -459,6 +459,24 @@ private function bail( string $message ) {
+			return;
+		}
+
+		if ( 'quirks' === $doctype->get_compatibility_mode() ) {
```
this reads a bit awkward to me. also one of the reasons I find direct property access helpful. the DOCTYPE token doesn't specify a document's compatibility mode - it can only suggest it when none is already provided or forced, right?

so we might say "this token indicates quirks mode" or "this token indicates no-quirks mode"

```php
if ( 'quirks' === $doctype->indicated_compatability_mode ) { }
```

but on this note do we want to constrain this function to only operate on a DOCTYPE token? I would imagine this could lead to people needing to try and push their own doctype declaration into existing HTML, when they could instead set this instead

```php
$processor->set_compatability_mode( WP_HTML_Tag_Processor::QUIRKS_MODE );
$processor->set_compatability_mode( WP_HTML_Tag_Processor::NO_QUIRKS_MODE );
```

Although this does in a sense expose an opportunity for a foot-gun, for someone to change the compatibility mode randomly, it doesn't force new rules onto the Tag Processor that I don't think fully exist there. we're not going to have a token in the fragment case you mention, so maybe we analyze the parsed DOCTYPE in the HTML Processor and then change modes in the Tag Processor via method call.

we can even `_doing_it_wrong()` if setting the compatibility mode after having already set it, which will send an error in debug modes but won't take down the site.

sirreal committed:
Refactor doctype info class to use from_html factory

Move the parsing of a complete DOCTYPE HTML token to the
::from_html factory method. This further reduces coupling between the
processor classes and the doctype info class.

sirreal committed:
Fix equals alignment lint

sirreal committed:
Make doctype info properties public and add more documentation

sirreal committed:
Update full parser compat mode from doctype handling

Remove the method to update the compat_mode and read the compat mode
property correctly.

sirreal committed:
Add readonly notes to doctype info properties

sirreal committed:
Add newline normalization and null byte replacement

sirreal committed:
Update tests to use direct property access

sirreal committed:
Fix off-by-one error on minimum length

sirreal committed:
Update missing doctype name test to use null

sirreal committed:
Move DOCTYPE tests to specific file

Simplify tag processor doctype tests

sirreal committed:
Fix lint

sirreal committed:
Merge branch 'trunk' into html-api/full-parser-doctype-quirks-mode-handling

sirreal committed:
Remove redundant extra argument in html5lib test helper

sirreal committed:
Check for undefined doctype identifiers in html5lib test trees

sirreal committed:
Remove test default argument that can't be used

sirreal committed:
Remove test arguments from removed dataProvider

sirreal committed:
Update doctype comments

Change the DOCTYPE status to suggest get_doctype_info over modifiable text.
It's a bit confusing because doctypes cannot set their "modifiable" text, which
makes the name modifiable awkward. It's unlikely this will be supported because
most docyptes are skipped, while other doctypes change how a document is parsed.

sirreal committed:
Final pass on documentation comments

dmsnell committed:
Documentating and naming updates.

dmsnell committed:
Add optimization for normative HTML DOCTYPE declaration.

dmsnell committed:
Merge remote-tracking branch 'upstream/trunk' into html-api/full-parser-doctype-quirks-mode-handling

dmsnell reviewed (approved):
@sirreal this is quality work of the level I've come to expect from you. thank you so much!

in my latest push I've made some renames and documentation updates, plus added a special-case for the normative `<!DOCTYPE html>`.

if you have any issues with these we can address them in a follow-up, but as we discussed out-of-band, I intend to merge this when the tests pass and when I'm able.

in follow-up work we can examine how the document compatability mode interact with the Tag Processor and CSS functions.

sirreal commented:
Thank you! I've reviewed your changes and I'm happy with them.

dmsnell committed:
Merge remote-tracking branch 'upstream/trunk' into html-api/full-parser-doctype-quirks-mode-handling

dmsnell commented:
Merged in [[58925]](https://core.trac.wordpress.org/changeset/58925)
https://github.com/wordpress/wordpress-develop/commit/1139a51e9c6608dc235938d9d234ed185c5d5706
